src/framebuffer/swtfb_client.rs

```diff
@@ -0,0 +1,224 @@
+
+        let sem_name_str = format!("/rm2fb.wait.{}", unsafe { libc::getpid() });
+        let mut sem_name = [0u8; 512];
+        for (i, byte) in sem_name_str.as_bytes().iter().enumerate() {
+            sem_name[i] = *byte;
+        }
+        self.send_wait_update(&wait_sem_data { sem_name });
+        let sem_name_c = CString::new(sem_name_str.as_str()).unwrap();
+        let sem = unsafe { libc::sem_open(sem_name_c.as_ptr() as *const u8, libc::O_CREAT) };
+
+        let mut timeout = libc::timespec {
+            tv_nsec: 0,
+            tv_sec: 0,
+        };
+        unsafe {
+            libc::clock_gettime(libc::CLOCK_REALTIME, &mut timeout);
+        }
+        timeout.tv_nsec += SEM_WAIT_TIMEOUT;
+
+        if timeout.tv_nsec >= 1_000_000_000 {
+            timeout.tv_nsec -= 1_000_000_000;
+            timeout.tv_sec += 1;
+        }
+        unsafe {
+            libc::sem_timedwait(sem, &timeout);
+            libc::sem_unlink(sem_name_c.as_ptr() as *const u8);
+        }
+    }
+
+    pub fn send_wait_update(&self, wait_update: &wait_sem_data) -> bool {
+        self.send(&swtfb_update {
+            mtype: MSG_TYPE::WAIT_t,
+            data: swtfb_update_data {
+                wait_update: *wait_update,
+            },
+        })
+    }
+
+    pub fn get_fix_screeninfo(&self) -> FixScreeninfo {
+        // https://github.com/ddvk/remarkable2-framebuffer/blob/1e288aa9/src/client/main.cpp#L217
+        let mut screeninfo: FixScreeninfo = unsafe { std::mem::zeroed() };
+        //screeninfo.smem_start = mem_map.as_ptr() as u32; // Not used anyway. TODO: Consider adding properly
+        screeninfo.smem_len = super::swtfb_client::BUF_SIZE as u32;
+        screeninfo.line_length =
+            super::swtfb_client::WIDTH as u32 * std::mem::size_of::<u16>() as u32;
+        screeninfo
+    }
+
+    pub fn get_var_screeninfo(&self) -> VarScreeninfo {
+        // https://github.com/ddvk/remarkable2-framebuffer/blob/1e288aa9/src/client/main.cpp#L194
+        let mut screeninfo: VarScreeninfo = unsafe { std::mem::zeroed() };
+        screeninfo.xres = super::swtfb_client::WIDTH as u32;
+        screeninfo.yres = super::swtfb_client::HEIGHT as u32;
+        screeninfo.grayscale = 0;
+        screeninfo.bits_per_pixel = 8 * std::mem::size_of::<u16>() as u32;
+        screeninfo.xres_virtual = super::swtfb_client::WIDTH as u32;
+        screeninfo.yres_virtual = super::swtfb_client::HEIGHT as u32;
+
+        //set to RGB565
+        screeninfo.red.offset = 11;
+        screeninfo.red.length = 5;
+        screeninfo.green.offset = 5;
+        screeninfo.green.length = 6;
+        screeninfo.blue.offset = 0;
+        screeninfo.blue.length = 5;
+        screeninfo
+    }
+}
+
+impl Drop for SwtfbClient {
+    fn drop(&mut self) {
+        if unsafe { libc::msgctl(self.msqid, libc::IPC_RMID, ptr::null_mut()) } != 0 {
+            panic!("Got an error when closing ipc queue!")
```

Comment:
[IIUC](https://doc.rust-lang.org/std/ops/trait.Drop.html#panics) panicking in `drop` will normally abort the process. Wondering if we should log here instead? IIUC this isn't a safety issue, so it might be more graceful to let the rest of shutdown proceed normally...

Reply:
Good point! I didn't even know about this. A warn! is probably best here.

Reply:
Should be fixed with 9343082.